### PR TITLE
[GPU] Fix dynamic concat optimization

### PR DESCRIPTION
**Pull request by p-durandin**
### Details:
 - Fix for dynamic layout in concat optimizations
 - Small fix for windows debug assert

### Tickets:
 - 112729

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 1]
```diff
@@ -143,7 +143,7 @@ bool concat_in_place_optimization::match(const program_node& concat_node,
             return false;
 
         layout pred_l = pred_params[idx].get_output_layout();
-        if (output_format != pred_l.format || output_datatype != pred_l.data_type)
+        if (output_format != pred_l.format || output_datatype != pred_l.data_type || pred_l.is_dynamic())
```
[the remark this answers is not in the document]
Comment (p-durandin) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:146]: It is failed in plugins\intel_gpu\src\runtime\layout.cpp:109 on is_dynamic() call

[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:146]: I wanted to ask where it is called from :)

[the remark this answers is not in the document]
Comment (p-durandin) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp]: @Lyamin-Roman proposed better solution, please review

[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp]: which one? There were multiple issues so I prepared PR https://github.com/openvinotoolkit/openvino/pull/17909

[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp]: The intention is "allow as much as possible at build time" and then disable optimization at runtime if not possible with the shape at that moment. So please see my change and do not apply current change of this PR

**Pull request closed without merging**